Documentation/upgrade.md

```diff
@@ -0,0 +1,293 @@
+---
+title: Upgrading Rook
+weight: 40
+---
+
+# Upgrading the Rook Software
+This guide will walk you through the manual steps to upgrade the software in a Rook cluster from one version to the next.
+Rook is a distributed software system and therefore there are multiple components to individually upgrade in the sequence defined in this guide.
+After each component is upgraded, it is important to verify that the cluster returns to a healthy and fully functional state.
+
+
+This guide is just the beginning of upgrade support in Rook.
+The goal is to provide prescriptive guidance and knowledge on how to upgrade a live Rook cluster and we hope to get valuable feedback from the community that will be incorporated into an automated upgrade solution by the Rook operator.
+
+We welcome feedback and opening issues!
+
+## Considerations
+With this manual upgrade guide, there are a few notes to consider:
+* **WARNING:** Upgrading a Rook cluster is a manual process in its very early stages.  There may be unexpected issues or obstacles that damage the integrity and health of your storage cluster, including data loss.  Only proceed with this guide if you are comfortable with that.
+* Rook is still in an alpha state.  Migrations and general support for breaking changes across versions are not supported or covered in this guide.
+
+## Prerequisites
+In order to successfully upgrade a Rook cluster, the following prerequisites must be met:
+* The cluster should be in a healthy state with full functionality.
+Review the [health verification section](#health-verification) in order to verify your cluster is in a good starting state.
+* `dataDirHostPath` must be set in your Cluster spec.
+This persists metadata on host nodes, enabling pods to be terminated during the upgrade and for new pods to be created in their place.
+More details about `dataDirHostPath` can be found in the [Cluster CRD readme](./cluster-crd.md#cluster-settings).
+
+The minimal sample Cluster spec that will be used in this guide can be found below (note that the specific configuration may not be applicable to all environments):
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: rook
+---
+apiVersion: rook.io/v1alpha1
+kind: Cluster
+metadata:
+  name: rook
+  namespace: rook
+spec:
+  versionTag: v0.5.0
```

Comment:
this versionTag will need to be updated in the cluster.yaml to v0.5.1. Otherwise, if you start rgw or mds services after this, they will get the v0.5.0 version from this tag.

Reply:
is that true?  My understanding is that in the API section (https://github.com/jbw976/rook/blob/upgrade-guide/Documentation/upgrade.md#api), we're telling the user to also update the `ROOK_VERSION_TAG` env var.  This is passed into the API daemon process (https://github.com/rook/rook/blob/master/cmd/rook/api.go#L43) and will get used by the k8s API handler (https://github.com/rook/rook/blob/master/cmd/rook/api.go#L77).

When launching RGW/MDS, the API handler uses that value, not the value from the Cluster CRD, right?
https://github.com/rook/rook/blob/master/pkg/api/k8s/cluster_handler.go#L69

Reply:
With #703 and #974 the operator will no longer be calling the api service to perform operations like start the services. The operator will perform these operations directly, thus the versionTag would be relevant from the clusterCRD

Reply:
sounds like this guide should indicate changing it in both locations then: in the cluster CRD and in the API deployment template.

When the version tag is updated in the cluster CRD, will the operator notice it since it doesn't handle update events to the cluster CRD currently?

Maybe the flow would be to first update the version tag on the cluster CRD, then update the operator itself.  The new operator pod that comes up would load the most up to date version of the Cluster CRD and have the most up to date version tag.  Sound correct?

Reply:
We need to update the version of the operator first, then the versionTag of the cluster. With the automated upgrade, this is how the operator will trigger it. Then if the operator decides special migration is needed during the upgrade, it will handle that.

Reply:
In these manual steps, with no automatic upgrade yet, and the operator not supporting cluster CRD updates, I think we need to update the cluster version tag first.  Otherwise, how would the operator ever find out about the cluster version tag getting updated?

Reply:
ah i see what you're saying. For these manual steps and in the current state of things agreed that the crd should be updated first.